client/rest-high-level/src/test/java/org/elasticsearch/client/RequestConvertersTests.java

```diff
@@ -2456,11 +2456,11 @@ public void testEndpointBuilderEncodeParts() {
         }
         {
             EndpointBuilder endpointBuilder = new EndpointBuilder().addPathPart("foo+bar");
-            assertEquals("/foo+bar", endpointBuilder.build());
+            assertEquals("/foo%2Bbar", endpointBuilder.build());
         }
         {
             EndpointBuilder endpointBuilder = new EndpointBuilder().addPathPart("foo+bar");
-            assertEquals("/foo+bar", endpointBuilder.build());
+            assertEquals("/foo%2Bbar", endpointBuilder.build());
```

Comment:
@javanna yea now that you mention it, I think the problem is with the code that handles the request, not the request itself.
`+` doesn't need to be encoded in a path component.

=> this PR and the issue seem to be missing the actual problem?

Reply:
A `+` does not need to be encoded by the specifications indeed. However, the issue here is that we treat `+` specially, converting it to a ` `. This is because some browsers encode spaces as `+`. See: https://github.com/elastic/elasticsearch/blob/f7a9186372edcd9da31a89ac0565d53874db1839/server/src/main/java/org/elasticsearch/rest/RestUtils.java#L166-L168

Reply:
I think the actual problem is that if you have an ID that contains a `+` and you try to use the HLRC to get a document by that then since we convert `+` to space server side then you can not get that document by ID unless you also encode the `+`.

Reply:
@jasontedor yes exactly that is the issue. So do we do the encoding here to account for the non-standard server behaviour?

Reply:
interesting, is this something that we want to fix/change on the server-side rather than adapting the client then?

Reply:
Yes, I think it is worth us making that change although it's a breaking one so we have to be careful and age it in  slowly, and maybe provide a system property for BWC for a time period.

Reply:
Section 2.4 of RFC 3986 is [germane](https://tools.ietf.org/html/rfc3986#section-2.4) here. Since we never treat `+` as a sub-component delimiter (unlike, e.g., `&`) we do not need to require that it be encoded when it is to be treated as data. Treating a plus as a space is only required by the specification in the query string component of `application/x-www-form-urlencoded` requests. Since we do not handle this media type for any requests with bodies anyway, we never need to treat `+` as ` `. Our handling here is just odd, I think it's safe for us to remove this legacy behavior.

Reply:
Im +1 to @jasontedor logic here.